src/BreakpointViewer.cpp

```diff
@@ -438,145 +477,165 @@ void BreakpointViewer::setBreakState()
 	runState = false;
 }
 
-void BreakpointViewer::sync()
+void BreakpointViewer::stretchTable(BreakpointData::Type type)
 {
-	auto sa = ScopedAssign(userMode, false);
-	keepUnusedItems();
+	if (type == BreakpointData::ALL) {
+		for (auto table: selector) {
+			// will not stretch without sorting
+			table->setSortingEnabled(true);
+			table->resizeColumnsToContents();
+		}
+		return;
+	}
 
-	int bpRow = bpTableWidget->rowCount();
-	int wpRow = wpTableWidget->rowCount();
-	int cnRow = cnTableWidget->rowCount();
+	auto table = selector[type];
+	// will not stretch without sorting
+	table->setSortingEnabled(true);
+	table->resizeColumnsToContents();
+}
 
-	if (breakpoints->breakpointCount() != -1) {
-		for (int index = 0; index < breakpoints->breakpointCount(); ++index) {
-			const Breakpoint& bp = breakpoints->getBreakpoint(index);
+QMap<QString, BreakpointData>::iterator
+BreakpointViewer::scanBreakpointData(const Breakpoint& bp)
+{
+	BreakpointData::Type type = bp.type == Breakpoint::BREAKPOINT ? BreakpointData::BREAKPOINT
+	                         : (bp.type == Breakpoint::CONDITION  ? BreakpointData::CONDITION : BreakpointData::WATCHPOINT);
+	auto map = MapList[type];
+	if (map->isEmpty()) return map->end();
+
+	// assumed premise: a new bp will never repeat same name again
+	if (QMap<QString, BreakpointData>::iterator it; (it = map->find(bp.id)) != map->end() && it->index == -1) {
+		// if name remains the same, bp has not changed.
+		return it;
+	}
 
-			switch (bp.type) {
-			case Breakpoint::BREAKPOINT: {
-				bpRow = createTableRow(BREAKPOINT);
-				setBreakpointStatus(BREAKPOINT, bpRow, true);
+	return map->end();
+}
 
-				bool currentSlot = breakpoints->inCurrentSlot(bp);
-				QBrush color = palette().brush(runState | currentSlot ?
-					QPalette::ColorGroup::Normal : QPalette::ColorGroup::Disabled,
-					QPalette::WindowText);
+void BreakpointViewer::connectBreakpoint(const QString& id, BreakpointData& data)
+{
+	MapList[data.type]->insert(id, data);
+}
 
-				QTableWidgetItem* item1 = bpTableWidget->item(bpRow, ENABLED);
-				item1->setCheckState(Qt::Checked);
+void BreakpointViewer::sync()
+{
+	// don't reload if generated its own update.
+	if (selfUpdating) {
+		selfUpdating = false;
+		return;
+	}
 
-				QTableWidgetItem* item2 = bpTableWidget->item(bpRow, BP_ADDRESS);
-				item2->setForeground(color);
-				item2->setText(hexValue(bp.address, 4));
+	bool empty[3] = {false, false, false};
+	int type = 0;
 
-				QTableWidgetItem* item3 = bpTableWidget->item(bpRow, T_CONDITION);
-				item3->setForeground(color);
-				item3->setText(bp.condition.trimmed());
+	for (auto map: MapList) {
```

Comment:
(detail, style)
My preference would be to write this as:
```for (auto* map : MapList) {```
So `auto*` instead of `auto`, and space at both sides of `:`.

This does not change the meaning of the code in any way. But personally I prefer to make the (deduced) pointer-type explicitly visible (and tools like `clang-tidy` agree with me). Rationale: `const` and `*` are automatically deduced as part of `auto`, but `&` is not (if you don't write `auto&` you get a copy instead of a reference). For this reason I prefer to always write all qualifiers.